.circleci/scripts/rerun-ci-workflow-from-failed.ts

```diff
@@ -0,0 +1,209 @@
+const CIRCLE_TOKEN = process.env.API_V2_TOKEN;
+
+interface Actor {
+  login: string;
+  avatar_url: string | null;
+}
+
+interface Trigger {
+  received_at: string;
+  type: string;
+  actor: Actor;
+}
+
+interface VCS {
+  origin_repository_url: string;
+  target_repository_url: string;
+  revision: string;
+  provider_name: string;
+  branch: string;
+}
+
+interface WorkflowItem {
+  id: string;
+  errors: string[];
+  project_slug: string;
+  updated_at: string;
+  number: number;
+  state: string;
+  created_at: string;
+  trigger: Trigger;
+  vcs: VCS;
+}
+
+interface CircleCIResponse {
+  next_page_token: string | null;
+  items: WorkflowItem[];
+}
+
+interface WorkflowStatusItem {
+  pipeline_id: string;
+  id: string;
+  name: string;
+  project_slug: string;
+  tag?: string;
+  status: string;
+  started_by: string;
+  pipeline_number: number;
+  created_at: string;
+  stopped_at: string;
+}
+
+interface WorkflowStatusResponse {
+  next_page_token: string | null;
+  items: WorkflowStatusItem[];
+}
+
+/**
+ * Fetches the last 20 CircleCI workflows for 'develop' branch.
+ * Note: the API returns the first 20 workflows by default.
+ * If we wanted to get older workflows, we would need to use the 'page-token' we would get in the first response
+ * and perform a subsequent request with the 'page-token' parameter.
+ * This seems unnecessary as of today, as the amount of daily PRs merged to develop is not that high.
+ *
+ * @returns {Promise<WorkflowItem[]>} A promise that resolves to an array of workflow items.
+ * @throws Will throw an error if the CircleCI token is not defined or if the HTTP request fails.
+ */
+async function getCircleCiWorkflowsByBranch(branch: string): Promise<WorkflowItem[]> {
+  if (!CIRCLE_TOKEN) {
+    throw new Error('CircleCI token is not defined');
+  }
+
+  const url = `https://circleci.com/api/v2/project/github/${process.env.CIRCLE_PROJECT_USERNAME}/${process.env.CIRCLE_PROJECT_REPONAME}/pipeline?branch=${branch}`;
+  const options = {
+    method: 'GET',
+    headers: {
+      'Content-Type': 'application/json',
+      'Circle-Token': CIRCLE_TOKEN,
```

Comment:
Optional: when i was consuming other apis from circle-ci, it seems like we don't need a token to access. Maybe this could be removed also, wdyt?

Reply:
hey @DDDDDanica , that's a great insight! I've just checked each request , and I see that the ones that are for reading workflows and pipelines (GET) are indeed available without a token, but the one for re-running a worflow (POST) does need a token.

I imagine with/without authentication can affect the rate limits, however I haven't found any information around this in [the Circle ci V2 API docs](https://circleci.com/docs/api/v2/index.html)

I'm happy with both options! Do you think removing it would be more advisable? 
Thank you in advance :pray: 

https://github.com/user-attachments/assets/b1f2b4aa-23aa-4c1e-a4ec-dbf59217b6ce

Reply:
I just found the rate limits, for both authenticated and non-authenticated seem to be 51 requests/second.

We could use the unauthenticated case for the GET requests, and the authenticated case for POST requests. Would that make sense?

In this job, whenever is triggered, we are making: 1+20+(failed jobs rerun) requests, which in the worst case scenario, we had to rerun all (very unlikely) we would be doing 41 requests ~second.

![Screenshot from 2024-11-15 12-51-52](https://github.com/user-attachments/assets/9992e7c3-daa0-4806-afdc-d8ae47741208)

![Screenshot from 2024-11-15 12-52-04](https://github.com/user-attachments/assets/b9dab9de-dadb-4985-b25e-ff9bda7d15b7)

Reply:
Sounds good Mariona, thanks for the research, it's good to know that there are some endpoints requiring the auth token. I would say whichever way to make the pipeline run faster and less costy is the way we go. The proposal looks good to me !